pennylane_lightning_kokkos/src/algorithms/ObservablesKokkos.hpp

```diff
@@ -0,0 +1,504 @@
+        buffer.initZeros();
+
+        for (size_t term_idx = 0; term_idx < coeffs_.size(); term_idx++) {
+            StateVectorKokkos<T> tmp(sv);
+            obs_[term_idx]->applyInPlace(tmp);
+            Util::axpy_Kokkos<T>(Kokkos::complex<T>{coeffs_[term_idx], 0.0},
+                                 tmp.getData(), buffer.getData(),
+                                 tmp.getLength());
+        }
+        sv.updateData(buffer);
+    }
+
+    [[nodiscard]] auto getWires() const -> std::vector<size_t> override {
+        std::unordered_set<size_t> wires;
+
+        for (const auto &ob : obs_) {
+            const auto ob_wires = ob->getWires();
+            wires.insert(ob_wires.begin(), ob_wires.end());
+        }
+        auto all_wires = std::vector<size_t>(wires.begin(), wires.end());
+        std::sort(all_wires.begin(), all_wires.end(), std::less{});
+        return all_wires;
+    }
+
+    [[nodiscard]] auto getObsName() const -> std::string override {
+
+        using Pennylane::Util::operator<<;
+        std::ostringstream ss;
+        ss << "Hamiltonian: { 'coeffs' : " << coeffs_ << ", 'observables' : [";
+        const auto term_size = coeffs_.size();
+        for (size_t t = 0; t < term_size; t++) {
+            ss << obs_[t]->getObsName();
+            if (t != term_size - 1) {
+                ss << ", ";
+            }
+        }
+        ss << "]}";
+        return ss.str();
+    }
+};
+
+/**
+ * @brief Sparse representation of HamiltonianKokkos<T>
+ *
+ * @tparam T Floating-point precision.
+ */
+template <typename T>
+class SparseHamiltonianKokkos final : public ObservableKokkos<T> {
+  public:
+    using PrecisionT = T;
+
+  private:
+    std::vector<std::complex<T>> data_;
+    std::vector<std::size_t> indices_; // colum indices
+    std::vector<std::size_t> indptr_;  // row_map
+    std::vector<std::size_t> wires_;
+
+    [[nodiscard]] bool
+    isEqual(const ObservableKokkos<T> &other) const override {
+        const auto &other_cast =
+            static_cast<const SparseHamiltonianKokkos<T> &>(other);
+
+        if (data_ != other_cast.data_ || indices_ != other_cast.indices_ ||
+            indptr_ != other_cast.indptr_) {
+            return false;
+        }
+
+        return true;
+    }
+
+  public:
+    /**
+     * @brief Create a SparseHamiltonian from data, indices and indptr in CSR
+     * format.
+     *
+     * @param arg1 Arguments to construct data
+     * @param arg2 Arguments to construct indices
+     * @param arg3 Arguments to construct indptr
+     * @param arg4 Arguments to construct wires
+     */
+    template <typename T1, typename T2, typename T3 = T2,
+              typename T4 = std::vector<std::size_t>>
+    SparseHamiltonianKokkos(T1 &&arg1, T2 &&arg2, T3 &&arg3, T4 &&arg4)
+        : data_{std::forward<T1>(arg1)}, indices_{std::forward<T2>(arg2)},
+          indptr_{std::forward<T3>(arg3)}, wires_{std::forward<T4>(arg4)} {
+        assert(data_.size() == indices_.size());
```

Comment:
Can this be a `PL_ASSERT` or equivalent to ensure it isn't removed for release builds?

Reply:
will do